### PR TITLE
Fixtimeouts

### PR DESCRIPTION
**Pull request by skirvan44**

### CODE REVIEW

**Review by brianjmurrell — APPROVED** [review]
LGTM but looks like you have a conflict that needs resolving.

**Review by phender — CHANGES_REQUESTED** [review]
The concern with just adjusting the avocado test timeout is it may conflict with other built-in or hard-coded timeouts.  For example nearly every test calls TestWithServers.setUp() which starts the agents and servers:
- Starting agents currently uses a [30 second](https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/agent_utils.py#L55) timeout
- Starting servers currently uses a hard coded [120 second](https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/server_utils.py#L469) timeout to perform the storage prepare prior to starting the servers, a [90 second](https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/server_utils.py#L45) timeout to start the daos_server command(s) and detect that the servers are ready, a [60 second](https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/command_utils.py#L661) timeout for the 'dmg storage format' command, and then the same [90 second](https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/server_utils.py#L45) timeout to detect that the daos_io_servers are running

With these excessive timeouts combined, any test with an avocado timeout less than 6.5 minutes has the possibility of timing out during setUp() and not leaving enough time to properly run all the tearDown() steps.

Now tearDown() has its own timeouts:
- if 'dmg' or 'daos' is being used to create the pools or containers created during the test, then each pool/container destroy is subject to a [60 second](https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/command_utils.py#L661) timeout
- stopping the agents/servers each uses a combination of timeouts to stop the orterun process in a loop that can run twice: [5 seconds](https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/command_utils.py#L214) to send a SIGTERM/SIGKILL and a (10 seconds)[https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/command_utils.py#L263] for each check to see if the daos commands are running remotely.

Assuming a test with one pool and one container, this could take 2.5 minutes at max timeouts.

The dmg/daos command [60 second](https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/command_utils.py#L661) will also affect any other dmg/daos commands being run in the test and this should probably be reduced to something more reasonable (5-10 seconds??) and increased in specific tests where it may be needed.

There are other timeouts to consider: 
- https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/general_utils.py#L75
- https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/ior_test_base.py#L411
- https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/ior_utils.py#L308

There are also some timeouts hard coded for various tests that should be verified that they are still valid (e.g. small enough to all the timeout to hit and provide enough time for tearDown() to complete) for any adjustments in the overall avocado test timeout.  One option here would be to change these hard coded values to test yaml driven values, allowing for easier future adjustment:
- https://github.com/daos-stack/daos/blob/master/src/tests/ftest/container/root_container_test.py#L256
-  https://github.com/daos-stack/daos/blob/master/src/tests/ftest/control/dmg_storage_reformat.py#L65
- https://github.com/daos-stack/daos/blob/master/src/tests/ftest/control/super_block_versioning.py#L58
- https://github.com/daos-stack/daos/blob/master/src/tests/ftest/io/dfuse_bash_cmd.py#L187
- https://github.com/daos-stack/daos/blob/master/src/tests/ftest/io/parallel_io.py#L227
-

**Comment by brianjmurrell** [comment]
> The concern with just adjusting the avocado test timeout is it may conflict with other built-in or hard-coded timeouts.

So I guess the question about all of those timeouts is are they grossly overestimated?  Surely they must be as we have tests that run and pass in much less than the worst-case accumulation of all of those timeouts, don't we?  Do we know the actual typical time for all of those operations?  Do we know why the timeouts are so high if they are grossly overestimated?

But on the other hand, even if they are grossly overestimated, they are really only a problem in the "perfect storm" situation, yes?  So only if things are really bad and go really wrong.  And even if they do, they are at least not leading to a test run taking 7-8 hours instead of 1-2 hours due to the compounding of overestimates over many variants.

**Comment by phender** [comment]
> > The concern with just adjusting the avocado test timeout is it may conflict with other built-in or hard-coded timeouts.
> 
> So I guess the question about all of those timeouts is are they grossly overestimated? Surely they must be as we have tests that run and pass in much less than the worst-case accumulation of all of those timeouts, don't we? Do we know the actual typical time for all of those operations? Do we know why the timeouts are so high if they are grossly overestimated?
> 
> But on the other hand, even if they are grossly overestimated, they are really only a problem in the "perfect storm" situation, yes? So only if things are really bad and go really wrong. And even if they do, they are at least not leading to a test run taking 7-8 hours instead of 1-2 hours due to the compounding of overestimates over many variants.

Yes they are probably too large with current behavior.  I believe after they were implemented start times went way down.  

I still think we should make a few adjustments to bring down the timeouts related to setUp(), dmg/daos commands, and tearDown() which should:
- get us around 60 seconds for a minimum avocado test timeout (that will still allow tearDown() to run)
- identify slowness in starting doas / running daos commands.  

These changes include (including examples of current execution times):
1. Reduce the default daos/dmg command timeout down from 60 seconds to 10 seconds (maybe even 5??): https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/command_utils.py#L661
```
    def __init__(self, namespace, command, path="", yaml_cfg=None, timeout=10):
```
```
2020-09-09 17:53:45,164 process          L0499 INFO | Command '/usr/bin/dmg -o /mnt/share/tmp.nfvzj8drL9/ior_small.py-1__daos_server_dmg.yaml pool create --group=jenkins --nsvc=1 --nvme-size=9000000000 --scm-size=3000000000 --sys=daos_server --user=jenkins' finished with 0 after 0.754907846451s
```
```
2020-09-09 17:53:45,402 process          L0499 INFO | Command '/usr/bin/daos container create --pool=06AB846F-345B-4847-9B16-8A73BC3EF240 --properties=cksum:crc16,cksum_size:16384,srv_cksum:on --svc=1 --sys-name=daos_server --type=POSIX' finished with 0 after 0.167599201202s
```
1a. There is one exception here, 'dmg storage format', which can be addressed by temporarily bumping up the timeout: https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/dmg_utils.py#L209
```
    def storage_format(self, reformat=False, timeout=20):
        """Get the result of the dmg storage format command.

        Args:
            reformat (bool): always reformat storage, could be destructive.
                This will create control-plane related metadata i.e. superblock
                file and reformat if the storage media is available and
                formattable.
            timeout (int): temporary timeout value to use with the storage
                format command

        Returns:
            CmdResult: an avocado CmdResult object containing the dmg command
                information, e.g. exit status, stdout, stderr, etc.

        Raises:
            CommandFailure: if the dmg storage format command fails.

        """
        restore_timeout = self.timeout
        self.timeout = timeout
        result = self._get_result(("storage", "format"), reformat=reformat)
        self.timeout = restore_timeout
        return result
```
```
2020-09-09 17:53:39,297 process          L0499 INFO | Command '/usr/bin/dmg -o /mnt/share/tmp.nfvzj8drL9/ior_small.py-1__daos_server_dmg.yaml storage format' finished with 0 after 12.3586189747s
```


2. Reduce the timeout for detecting agent startup to 5 seconds (seeing sub 1 second detection times): https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/agent_utils.py#L55
```
    def __init__(self, path="", yaml_cfg=None, timeout=5):
```
```
2020-09-09 17:53:03,753 command_utils    L0650 INFO | daos_agent subprocess startup detected - 5/5 'Listening on ' messages detected in 0.652227878571/30 seconds
```

3. Reduce the timeout for detecting server state chages to 15 seconds: https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/server_utils.py#L45
```
    def __init__(self, path="", yaml_cfg=None, timeout=15):
```
```
2020-09-09 17:53:26,914 command_utils    L0650 INFO | daos_server subprocess startup detected - 4/4 '(SCM format required)(?!;)' messages detected in 10.5630061626/90 seconds
```
```
2020-09-09 17:53:44,364 command_utils    L0650 INFO | daos_server subprocess startup detected - 4/4 'DAOS I/O server.*started' messages detected in 5.06474494934/90 seconds
```

4. Reduce the timeout for storage prepare prior to starting servers to 20-30 seconds: https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/server_utils.py#L469
```
        result = pcmd(self._hosts, str(cmd), timeout=20)
```
```
2020-09-09 17:53:04,643 server_utils     L0468 INFO | Preparing DAOS server storage: /usr/bin/daos_server storage prepare --force --hugepages=4096 --target-user=root
2020-09-09 17:53:16,304 output           L0655 DEBUG| Command:
2020-09-09 17:53:16,304 output           L0655 DEBUG|   /usr/bin/daos_server storage prepare --force --hugepages=4096 --target-user=root
2020-09-09 17:53:16,305 output           L0655 DEBUG| Command return codes:
2020-09-09 17:53:16,305 output           L0655 DEBUG|   wolf-[142-145]: rc=0
2020-09-09 17:53:16,305 output           L0655 DEBUG| Command output:
2020-09-09 17:53:16,306 output           L0655 DEBUG|   wolf-[142-145]:
2020-09-09 17:53:16,306 output           L0655 DEBUG|     Preparing locally-attached NVMe storage...
2020-09-09 17:53:16,306 output           L0655 DEBUG|     Preparing locally-attached SCM...
2020-09-09 17:53:16,306 output           L0655 DEBUG|     SCM namespaces:
2020-09-09 17:53:16,306 output           L0655 DEBUG|     			Device:pmem0 Socket:0 Capacity:799 GB
2020-09-09 17:53:16,306 output           L0655 DEBUG|     		Device:pmem1 Socket:1 Capacity:799 GB
```

The numbers above are based on passing observations of agent/server start times, so feel free to adjust as you see fit.  The above examples are from https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/2821/artifact/Functional/job-2020-09-09T17.53-ab3c0a0-io-ior_small/job.log.

If we have a way for searching all of the recent https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/28*/artifact/Functional/job-*/job.log files to obtain an average/max for the above execution times that would be ideal.

**Comment by skirvan44** [comment]
Changes made per Phil's comment.  We'll see if they work in CI.

File: src/tests/ftest/util/dmg_utils.py [hunk 1]
```diff
@@ -206,7 +206,7 @@ def storage_scan(self, verbose=False):
 
         return data
 
-    def storage_format(self, reformat=False):
+    def storage_format(self, reformat=False, timeout=20):
```
Comment (brianjmurrell) [inline src/tests/ftest/util/dmg_utils.py:209]: Does the new argument need to be added to the `Args:` description in the comment below?

**Review by brianjmurrell — COMMENTED** [review]
The default timeout (unless it's been changed since I set it a long time ago when I created apricot) is 60 seconds.  I wonder if we should manage all of the sub-60s tests getting the proposed 60s minimum by just omitting their timeout in the .yaml file.  That way if we need to increase the minimum, we can do it in one place, in apricot and all of those tests that are just getting the minimum automatically get the new default set in apricot.

File: src/tests/ftest/util/dmg_utils.py (hunk 1 above)
Comment (skirvan44) [reply src/tests/ftest/util/dmg_utils.py:209]: I'll fix the description.  Personally I like the explicit timeouts.

**Review by phender — CHANGES_REQUESTED** [review]
I missed two locations where defaults are used to set the timeouts for detecting agent and server startup:
- https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/apricot/apricot/test.py#L473
- https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/apricot/apricot/test.py#L500

These should be set to match the changes to default timeout values in `DaosAgentCommand` and `DaosServerCommand` - `5` and `15` respectively.

This `15` second default `timeout` value for `DaosServerCommand.__init__()` (util/server_utils.py) and (as mentioned above) for `TestWithServers.add_server_manager()` (util/apricot/apricot/test.py) should probably be increased too as the https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3377/4/artifact/Functional/job-2020-09-15T15.30-cde6e09-io-hdf5/job.log would have timed out if it was using 15 seconds:
```
2020-09-15 15:30:59,921 command_utils    L0650 INFO | daos_server subprocess startup detected - 1/1 '(SCM format required)(?!;)' messages detected in 15.2907149792/90 seconds
```

**Comment by phender** [comment]
Some observations from https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3377/4/artifact/Functional/job-2020-09-15T15.30-cde6e09-io-hdf5/job.log:

- 15:30:26,018 (+00:00) Test starts w/ a 60 second avocado test timeout
- 15:30:27,756 (+00:01)  daos_agent starts in 0.675898075104 seconds
- 15:30:28,609 (+00:01)  daos_server storage format issued
- 15:30:44,625 (+00:16)  daos_server start command issued
- 15:30:59,921 (+00:15)  daos_server "SCM format required" messages detected in 15.2907149792 seconds
- 15:30:59,957 (+00:00)  dmg storage format issued
- 15:31:16,036 (+00:17)  dmg storage format finished after 16.075660944s
- 15:31:18,438 (+00:02)  daos_io_server start messages detected in 2.39928197861 seconds
- 15:31:18,439 (+00:00) test method execution begins
- 15:31:56,462 (+00:38) avocado test timeout is triggered
- 15:31:56,504 (+00:00) tearDown starts
- 15:31:57,462 (+00:01) avocado stops test before tearDown completes

The overall <TestWithServers>.startUp() runtime in this case was 51 seconds (w/o hitting any timeouts), leaving only 38 seconds for the actual test to run (TestWithServers added a fixed 30 seconds to the new 60 second timeout: https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/apricot/apricot/test.py#L254), and really no time for tearDown() to execute - which is not good.

In https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/2837/artifact/Functional/job-2020-09-15T16.52-7c9d8b1-io-hdf5/job.log - where the test passed - we see:

- 16:52:11,106 (+00:00)  Test starts w/ a 1800 second avocado test timeout
- 16:52:12,419 (+00:01)  daos_agent starts in 0.642510890961 seconds
- 16:52:13,242 (+00:01)  daos_server storage format issued
- 16:52:24,925 (+00:11)  daos_server start command issued
- 16:52:35,253 (+00:11)  daos_server "SCM format required" messages detected in 10.3226850033 seconds
- 16:52:35,256 (+00:00)  dmg storage format issued
- 16:52:47,600 (+00:12)  dmg storage format finished after 12.3414309025s
- 16:52:50,072 (+00:03)  daos_io_server start messages detected in 2.46913599968 seconds
- 16:52:50,073 (+00:00) test method execution begins
- 16:54:32,552 (+01:42) test method completes
- 16:54:32,553 (+00:00) tearDown starts
- 16:54:48,564 (+00:16) tearDown completes

Ideally we should be breaking the avocado test timeout up into setup_timeout (default 60s) + test_timeout (default ??) + teardown_timeout (default 30s).  The issue is how to enforce the test_timeout so that is stops the 'test_*' method execution and leaves enough time (teardown_timeout) remaining in the avocado test timeout for tearDown to run.  This is relatively easy with tests using mpirun/orterun to execute some command - we can apply the timeout to mpirun/orterun - but it may not be as simple with other tests.

All that said 60 seconds is too short a timeout for the io/hdf5.py test.  It looks like it should be closer to 120 seconds (assuming TestWithServers retains adding 30 seconds to the avocado test timeout - which admittedly doesn't really accomplish what it is intended to do)

**Comment by phender** [comment]
> Ideally we should be breaking the avocado test timeout up into setup_timeout (default 60s) + test_timeout (default ??) + teardown_timeout (default 30s). The issue is how to enforce the test_timeout so that is stops the 'test_*' method execution and leaves enough time (teardown_timeout) remaining in the avocado test timeout for tearDown to run. This is relatively easy with tests using mpirun/orterun to execute some command - we can apply the timeout to mpirun/orterun - but it may not be as simple with other tests.

I feel like this work may be slightly beyond the scope of this PR - its not like the limited amount of time tearDown() has to run when the test times out is a new problem.  There are some changes coming in [PR 3318](https://github.com/daos-stack/daos/pull/3318) that will [enable better teardown handling](https://github.com/daos-stack/daos/blob/60feaca571b04d842d250947b984508b68c6d446/src/tests/ftest/util/apricot/apricot/test.py#L646) of mpirun/orterun (JobManager-based classes).  As an extension of this work a yaml-based 'job_manager_timeout' value could be added to help facilitate a more finite test_* method run time for tests that use mpirun/orterun as their main component.  This would be best addressed after 3318 is merged.

**Comment by phender** [comment]
As another option - again maybe after this lands - we could make use of the [timeout-decorator](https://pypi.org/project/timeout-decorator/) on each test_* method.

**Review by brianjmurrell — APPROVED** [review]

**Review by brianjmurrell — APPROVED** [review]

**Review by phender — APPROVED** [review]

**Pull request merged**